src/telegram_to_engagement_db/telegram_group_to_engagement_db.py

```diff
@@ -0,0 +1,240 @@
+
+from telethon import TelegramClient
+from telethon.errors import SessionPasswordNeededError
+from telethon.tl.types import (PeerChannel)
+
+from storage.google_cloud import google_cloud_utils
+from core_data_modules.logging import Logger
+
+from engagement_database.data_models import (Message, MessageDirections, MessageOrigin, MessageStatuses,
+                                             HistoryEntryOrigin)
+
+from src.telegram_to_engagement_db.cache import TelegramGroupSyncCache
+
+log = Logger(__name__)
+
+# TODO: move to social media tools
+async def _initialize_telegram_client(telegram_token_file_url, google_cloud_credentials_file_path, pipeline_name):
+    """
+    :param telegram_token_file_url: Path to the Google Cloud file path that contains telegram tokens api_id
+                                                api_hash and telegram app admin phone number.
+    :type telegram_token_file_url: str
+    :param google_cloud_credentials_file_path: Path to the Google Cloud service account credentials file to use when
+                                               downloading facebook page token.
+    :type google_cloud_credentials_file_path: str
+    :param pipeline: Name of the pipeline for this telegram session.
+    :type pipeline: str
+    :return telegram_client
+    :rtype: telethon.client.telegramclient.TelegramClient
+
+    """
+    log.info('Downloading telegram access tokens...')
+    telegram_tokens = json.loads(google_cloud_utils.download_blob_to_string(
+        google_cloud_credentials_file_path, telegram_token_file_url).strip())
+
+    api_id = telegram_tokens['api_id']
+    api_hash = telegram_tokens['api_hash']
+    phone_number = telegram_tokens['phone_number']
+
+    log.info('Initializing telegram client...')
+    telegram = TelegramClient(f'{pipeline_name}_telegram_session_name', api_id, api_hash)
+    await telegram.start()
+
+    # Ensure the client is authorized
+    # While authenticating the first time the client will send an auth code to the application phone number.
+    # The authentication details are stored in a session file to enable you authenticate automatically the next time the script runs.
+    # To re-authenticate delete the session file.
+    if not await telegram.is_user_authorized():
+        await telegram.send_code_request(phone_number)
+        try:
+            await telegram.sign_in(phone_number, input(f"Enter the authorization code sent to "
+                                                              f"telegram a/c for {phone_number}: "))
+        # If the a/c has 2FA enabled sign_in() will raise a SessionPasswordNeededError.
+        # Input your telegram password to proceed.
+        except SessionPasswordNeededError:
+            await telegram.sign_in(password=input('Password: '))
+    log.info('Initialized telegram client...')
+
+    return telegram
+
+
+async def _fetch_message_from_group(telegram, group_id, dataset_end_date=None, dataset_group_latest_seen_message_id=None):
```

Comment:
```suggestion
async def _fetch_message_from_group(telegram, group_id, dataset_end_date=None, dataset_group_start_message_id=None):
```
Suggesting this because when fetching I think all you care about is the data range, so using "latest_seen" in this case is a bit of a complexity this function doesn't need to worry about

Reply:
I disagree,  dataset_group_latest_seen_message_id best captures that this number is mutable with every run. dataset_group_start_message_id points to a predefined immutable id where each run will always start from.